typing-tutor/main.js

```diff
@@ -2,24 +2,35 @@
 
 var sentence = 'grumpy wizards make toxic brew for the evil queen and jack.'
 
-var characters = sentence.split('').map(function (char, index) {
+var characters = sentence.split('').map(function (char, index, failures) {
   return {
     char,
-    index
+    index,
+    failures: null
   }
 })
 
 var appState = {
-  character: characters,
+  characters: characters,
   currentCharacter: 0
 }
 
+var pressedKey
+
 function render(char) {
   var $char = document.createElement('span')
   $char.textContent = char.char
   if (char.index === appState.currentCharacter) {
     $char.classList.add('current-character')
   }
+  if (characters[appState.currentCharacter].failures > 0 && char.index === appState.currentCharacter) {
+    $char.className = 'failed'
+    appState.currentCharacter++
```

Comment:
Instead of updating `appState` here, you should do it in your event listener. `render` should not change anything, just compute a new `$char`.